### PR TITLE
Add a match state filter

### PR DESCRIPTION
**Pull request by KingOfSquares**
~~Sorry I didn't test I wanted to get this in time for Hacktoberfest, will of course test before its merged~~
It already ended 🤦 

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>

### CODE REVIEW

**Comment by Pugzy** [comment]
Proto 1.4.2 added filters that allow you to check the match state.

These can be found under the [Matcher Elements](http://docs.oc.tc/modules/filters) section and are `<match-started/>`, `<match-running/>` and `<match-finished/>`.

Not sure why one implementation would be favoured over another (others may able to comment) but you should probs add the ability to define them in that way to bring us closer to that sweet sweet 1.4.2 support.

**Comment by Pablete1234** [comment]
https://github.com/OvercastNetwork/ProjectAres/blob/master/PGM/src/main/java/tc/oc/pgm/filters/parser/FilterDefinitionParser.java#L463
I belive we should try to re-implement what's missing in a way that is compatible with what existed before

File: core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.util.function.Predicate;
+import tc.oc.pgm.api.filter.query.MatchQuery;
+import tc.oc.pgm.api.match.Match;
+
+public class MatchStateFilter extends TypedFilter<MatchQuery> {
+
+  private final Predicate<Match> matchPredicate;
+
+  public MatchStateFilter(Predicate<Match> matchPredicate) {
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java]: I don't get the obsession for a predicate here. Just use a `Set<MatchPhase>`

**Review by Pablete1234 — CHANGES_REQUESTED** [review]

File: core/src/main/java/tc/oc/pgm/filters/FilterParser.java [hunk 2]
```diff
@@ -443,4 +446,52 @@ public StructuralLoadFilter parseStructuralLoad(Element el) throws InvalidXMLExc
...
+        break;
+      case "before":
+        matchPredicate = m -> m.getPhase() == MatchPhase.IDLE;
+        break;
+      case "started":
+        matchPredicate =
+            m -> m.getPhase() == MatchPhase.RUNNING || m.getPhase() == MatchPhase.FINISHED;
+    }
+    if (matchPredicate == null)
+      throw new InvalidXMLException("Invalid or no match state found", el);
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/filters/FilterParser.java]: Seems like this should be a private function `parseMatchPhase(String) -> MatchPhase`

File: core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java (hunk 1 above)
Comment (Electroid) [reply core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java]: See my other comment, this constructor can then be `MatchPhaseFilter(MatchPhase)`

**Review by Electroid — CHANGES_REQUESTED** [review]

File: core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java (hunk 1 above)
Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/filters/MatchStateFilter.java]: Yes! Predicate was an old thing before i actually used all MatchPhases (was using method refrences instead like Match::isRunning etc.)

File: core/src/main/java/tc/oc/pgm/filters/FilterParser.java [hunk 3]
```diff
@@ -443,4 +442,52 @@ public StructuralLoadFilter parseStructuralLoad(Element el) throws InvalidXMLExc
...
+  public MatchPhaseFilter parseMatchRunning(Element el) throws InvalidXMLException {
+    return parseMatchPhaseFilter("running", el);
+  }
+
+  @MethodParser("match-finished")
+  public MatchPhaseFilter parseMatchFinished(Element el) throws InvalidXMLException {
+    return parseMatchPhaseFilter("finished", el);
+  }
+
+  private MatchPhaseFilter parseMatchPhaseFilter(String matchState, Element el)
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/filters/FilterParser.java]: I have an idea to make this easier:

* Change the signature of `parseMatchPhaseFilter` return a generic `Filter`
* Change `MatchPhaseFilter` to only accept 1 `MatchPhase` and not a `Set`
* For the "started" case, return an `OrFilter(MatchPhaseFilter(RUNNING), MatchPhaseFilter(FINISHED))`

See what I'm getting at?

**Review by Electroid — CHANGES_REQUESTED** [review]

File: core/src/main/java/tc/oc/pgm/filters/FilterParser.java (hunk 3 above)
Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/filters/FilterParser.java]: Yes! Great idea 👍

**Review by Electroid — DISMISSED** [review]

File: core/src/main/java/tc/oc/pgm/filters/FilterParser.java [hunk 4]
```diff
@@ -453,4 +454,50 @@ public TimeFilter parseTimeFilter(Element el) throws InvalidXMLException {
...
+        break;
+      case "starting":
+        matchPhase = MatchPhase.STARTING;
+        break;
+      case "idle":
+        matchPhase = MatchPhase.IDLE;
+        break;
+      case "started":
+        return AnyFilter.of(
+            new MatchPhaseFilter(MatchPhase.RUNNING), new MatchPhaseFilter(MatchPhase.FINISHED));
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/filters/FilterParser.java]:
```suggestion
      case "running": return MatchPhaseFilter.RUNNING;
      case "finished": return MatchPhaseFilter.FINISHED;
    // ...
```
As suggested in discord you should make static final fields in MatchPhaseFilter and reuse the same instances

**Review by Pablete1234 — CHANGES_REQUESTED** [review]

**Review by Pablete1234 — APPROVED** [review]

**Comment by Pablete1234** [comment]
@Electroid could we get this merged?

**Review by Electroid — APPROVED** [review]

**Pull request merged**